### PR TITLE
feat: add `data-component` attribute to DOM elements

### PR DESCRIPTION
**Pull request by roshaans**
demo shows the component you interacted with in the console window. It is using the forked VM (implements the data-component attribute which shows the component source responsible for rendering the DOM element. 

demo: https://near-discovery-pvw23dttv-near-developer-console.vercel.app/?

- Enables easier component discovery for devs
- makes it possible to get insights on what components users are using. The data collected from the gateway will be made available to everyone. 

demo vid: https://www.loom.com/share/7c76eb265b6a435e92e3de021ffa742d

### How to use this in your gateway?

In your gateway pass a `features` object with `enableComponentSrcDataKey: true`

You would add this in `viewer` at this [line](https://github.com/NearSocial/viewer/blob/cfde8d4ae6504965933fe798a15584a9420f9cda/src/App.js#L88). 

```
      initNear({
        networkId: NetworkId,
        customElements: {}
        config: {
          defaultFinality: undefined,
        },
        features: {
          "enableComponentSrcDataKey": false,
        }
      });
```

### CODE REVIEW

**Comment by evgenykuzyakov** [comment]
I would still prefer to make it optional and not enabled by default. We should add feature flags to the VM, to enable certain optional actions.

**Comment by roshaans** [comment]
> I would still prefer to make it optional and not enabled by default. We should add feature flags to the VM, to enable certain optional actions.

Would you prefer a simple `featureFlagsMap` to be passed to the VM or create a `FeatureFlags` class to be created?

This class will make sure we keep track of all new feature flags in one place and that there is a unified way to check if a feature is enabled or not.

This is what the interface would roughly look like, but it'd be in Javascript

```javascript
interface IFeatureFlags {
  isFeatureEnabled(featureName: string): boolean;
  enableFeature(featureName: string): void;
  disableFeature(featureName: string): void;
  setMultipleFeatures(featuresMap: Map<string, boolean>): void;
  getAllFeatures(): { featureName: string, isEnabled: boolean }[];
}

class FeatureFlags implements IFeatureFlags {
  private features: Map<string, boolean>;

  constructor() {
    this.features = new Map([
      ['addWidgetSrcAttribute', false]
// other features as we add them
    ]);
  }

  // ... (rest of the methods)
}
```

**Comment by evgenykuzyakov** [comment]
A simple map should be fine. E.g. an object `features`, with attributes like `enableWidgetSrcDataKey: true`

File: src/lib/vm/vm.js [hunk 1]
```diff
@@ -586,6 +586,11 @@ class VmStack {
     });
     attributes.key =
       attributes.key ?? `${this.vm.widgetSrc}-${element}-${this.vm.gIndex}`;
+
+    if (this.vm.near.features.enableWidgetSrcDataKey == true) {
```
Comment (evgenykuzyakov) [inline src/lib/vm/vm.js:590]: Probably safer.

```suggestion
    if (this.vm.near?.features?.enableWidgetSrcDataKey == true) {
```

File: src/lib/data/near.js [hunk 2]
```diff
@@ -218,6 +218,7 @@ async function _initNear({
   selector,
   walletConnectCallback = () => {},
   customElements = {},
+  features = {},
```
Comment (evgenykuzyakov) [inline src/lib/data/near.js:221]: Let's add a comment above the `_initNear` that defines which features are available to keep track in one place. With description of a feature.

**Review by evgenykuzyakov — CHANGES_REQUESTED** [review]
Please rebase to `dev` and add changes to CHANGELOG.md

**Review by evgenykuzyakov — APPROVED** [review]

File: src/lib/vm/vm.js [hunk 3]
```diff
@@ -583,6 +583,11 @@ class VmStack {
     });
     attributes.key =
       attributes.key ?? `${this.vm.widgetSrc}-${element}-${this.vm.gIndex}`;
+
+    if (this.vm.near?.features?.enableWidgetSrcDataKey == true) {
+      attributes["data-key"] = attributes["data-key"] ?? this.vm.widgetSrc;
```
Comment (evgenykuzyakov) [inline src/lib/vm/vm.js:588]:
```suggestion
      attributes["data-widgetSrc"] = attributes["data-widgetSrc"] ?? this.vm.widgetSrc;
```

File: src/lib/data/near.js [hunk 4]
```diff
@@ -211,13 +211,18 @@ async function web4ViewCall(contractId, methodName, args, fallback) {
   }
 }
 
+/**
+ * Current VM Features:
+ * - enableWidgetSrcDataKey: Allows enabling the widget source data-key for rendered DOM elements. Disabled by default.
```
Comment (evgenykuzyakov) [inline src/lib/data/near.js:216]:
```suggestion
 * - enableWidgetSrcDataKey: Allows enabling the widget source data-widgetSrc for rendered DOM elements. Disabled by default.
```

**Review by evgenykuzyakov — CHANGES_REQUESTED** [review]
Please add changes to `CHANGELOG.md`

**Review by evgenykuzyakov — APPROVED** [review]

**Comment by alsakhaev** [comment]
Do we really need to add a `data-component' attribute to every DOM node? Isn't just the root node enough?
Check out my commit [here] (https://github.com/dapplets/near-social-vm/commit/f9008dd6ddac2402f6bc3095dcfe0a00c14dbcf1)

Also I ran into the React warning: 
```
Warning: Invalid prop `data-component` passed to `React.Fragment`. React.Fragment can only have key and child props.
```
We should check that a node is not a React.Fragment before adding the attribute. I did it with the `isReactObject(result)` function.

**Pull request merged**